src/Umbraco.Web.Common/Extensions/OAuthOptionsExtensions.cs

```diff
@@ -0,0 +1,73 @@
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Umbraco.Cms.Core.Configuration.Models;
+using Umbraco.Cms.Core.DependencyInjection;
+
+namespace Umbraco.Cms.Web.Common.Extensions;
+
+public static class OAuthOptionsExtensions
+{
+    /// <summary>
+    /// Applies SetUmbracoRedirectWithFilteredParams to both OnAccessDenied and OnRemoteFailure
+    /// on the OAuthOptions so Umbraco can do its best to nicely display the error messages
+    /// that are passed back from the external login provider on failure.
+    /// </summary>
+    /// <param name="oAuthOptions"></param>
+    /// <typeparam name="T"></typeparam>
+    /// <returns></returns>
+    public static T SetDefaultErrorEventHandling<T>(this T oAuthOptions) where T : OAuthOptions
+    {
+        oAuthOptions.Events.OnAccessDenied = HandleResponseWithDefaultUmbracoRedirect;
+        oAuthOptions.Events.OnRemoteFailure = HandleResponseWithDefaultUmbracoRedirect;
+
+        return oAuthOptions;
+    }
+
+    private static Task HandleResponseWithDefaultUmbracoRedirect(HandleRequestContext<RemoteAuthenticationOptions> context)
+    {
+        context.SetUmbracoRedirectWithFilteredParams()
+            .HandleResponse();
+
+        return Task.FromResult(0);
+    }
+
+    /// <summary>
+    /// Sets the context to redirect to the <see cref="SecuritySettings.AuthorizeCallbackErrorPathName"/> path with all parameters, except state, that are passed to the initial server callback configured for the configured external login provider
+    /// </summary>
+    /// <returns></returns>
+    public static T SetUmbracoRedirectWithFilteredParams<T>(this T context)
+        where T : HandleRequestContext<RemoteAuthenticationOptions>
+    {
+        var callbackPath = StaticServiceProvider.Instance.GetRequiredService<IOptions<SecuritySettings>>().Value
+            .AuthorizeCallbackErrorPathName;
+        context.Response.Redirect($"{callbackPath}?" + string.Join(
```

Comment:
We have a string extension to append a query string: `String.AppendQueryStringToUrl` that you could consider to use